src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs

```diff
@@ -545,14 +633,73 @@ private static bool CanUseScatterGatherWindowsAPIs(SafeFileHandle handle)
+
             // For pipes the offsets are ignored by the OS
-            nativeOverlapped.OffsetLow = unchecked((int)fileOffset);
-            nativeOverlapped.OffsetHigh = (int)(fileOffset >> 32);
+            result->OffsetLow = unchecked((int)fileOffset);
+            result->OffsetHigh = (int)(fileOffset >> 32);
+
+            // From https://docs.microsoft.com/en-us/windows/win32/api/ioapiset/nf-ioapiset-getoverlappedresult:
+            // "If the hEvent member of the OVERLAPPED structure is NULL, the system uses the state of the hFile handle to signal when the operation has been completed.
+            // Use of file, named pipe, or communications-device handles for this purpose is discouraged.
+            // It is safer to use an event object because of the confusion that can occur when multiple simultaneous overlapped operations
+            // are performed on the same file, named pipe, or communications device.
+            // In this situation, there is no way to know which operation caused the object's state to be signaled."
+            // Since we want RandomAccess APIs to be thread-safe, we provide a dedicated wait handle.
+            result->EventHandle = resetEvent.SafeWaitHandle.DangerousGetHandle();
+
+            return result;
+        }
 
-            return nativeOverlapped;
+        private static NativeOverlapped GetNativeOverlappedForSyncHandle(SafeFileHandle handle, long fileOffset)
+        {
+            Debug.Assert(!handle.IsAsync);
+
+            NativeOverlapped result = default;
+            result.OffsetLow = unchecked((int)fileOffset);
+            result.OffsetHigh = (int)(fileOffset >> 32);
+            return result;
+        }
+
+        private static unsafe IOCompletionCallback AllocateCallback()
+        {
+            return new IOCompletionCallback(Callback);
+
+            static unsafe void Callback(uint errorCode, uint numBytes, NativeOverlapped* pOverlapped)
+            {
+                CallbackResetEvent state = (CallbackResetEvent)ThreadPoolBoundHandle.GetNativeOverlappedState(pOverlapped)!;
+                state.FreeNativeOverlapped(pOverlapped);
+            }
+        }
+
+        // We need to store the reference count (see the comment in FreeNativeOverlappedIfItIsSafe) and an EventHandle to signal the completition.
+        // We could keep these two things separate, but since ManualResetEvent is sealed and we want to avoid any extra allocations, this type has been created.
+        // It's basically ManualResetEvent with reference count.
+        private sealed class CallbackResetEvent : EventWaitHandle
+        {
+            private int _freeWhenZero = 2; // one for the callback and another for the method that calls GetOverlappedResult
+            private ThreadPoolBoundHandle _threadPoolBoundHandle;
+
+            internal CallbackResetEvent(bool initialState, ThreadPoolBoundHandle threadPoolBoundHandle) : base(initialState, EventResetMode.ManualReset)
```

Comment:
We never pass true to initialState (and there'd be no reason for us to).  We can just remove it from the ctor and explicitly pass `initialState: false` to the base.